core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -3029,6 +3069,7 @@ class ReplicaManager(val config: KafkaConfig,
       stateChangeLogger.info(s"Started fetchers as part of become-follower for ${partitionsToStartFetching.size} partitions")
 
       partitionsToStartFetching.keySet.foreach(completeDelayedFetchOrProduceRequests)
+      partitionsToStartFetching.keySet.foreach(completeDelayedShareFetchRequest)
```

Comment:
1. It seems that all existing callers to `completeDelayedFetchOrProduceRequests` need to also check the share purgatory. Perhaps we could just add the `completeDelayedShareFetchRequest` call inside `completeDelayedFetchOrProduceRequests`.
2. In the case when a partition is deleted or becomes a follower, we probably also want to remove the sharePartition from `SharePartitionManager` to free up the space.

Reply:
1. I have added the change in my current PR.
2. I have created a JIRA https://issues.apache.org/jira/browse/KAFKA-17783 for this, we'll take it up in the future PRs.